recipe/patches/0005-CMake-Explicitly-link-with-CCCL-standalone-or-CTK.patch

```diff
@@ -0,0 +1,86 @@
+From: Hyunsu Cho <phcho@nvidia.com>
+Date: Wed, 24 Jul 2024 11:00:56 -0700
+Subject: [PATCH] [CMake] Explicitly link with CCCL (standalone or CTK)
+
+Also allow building with CCCL that's newer than CTK
+
+Co-authored-by: jakirkham <jakirkham@gmail.com>
+---
+ CMakeLists.txt        | 18 ++++++++++++++++++
+ cmake/Utils.cmake     |  7 ++++---
+ src/common/quantile.h |  8 ++++++++
+ 3 files changed, 30 insertions(+), 3 deletions(-)
+
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index dc0f7e654..aa078405f 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -224,6 +224,24 @@ if(USE_CUDA)
+   add_subdirectory(${PROJECT_SOURCE_DIR}/gputreeshap)
+ 
+   find_package(CUDAToolkit REQUIRED)
++  find_package(CCCL CONFIG)
++  if(NOT CCCL_FOUND)
++    message(STATUS "Standalone CCCL not found. Attempting to use CCCL from CUDA Toolkit...")
++    find_package(CCCL CONFIG
++      HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++    if(NOT CCCL_FOUND)
++      message(STATUS "Could not locate CCCL from CUDA Toolkit. Using Thrust and CUB from CUDA Toolkit...")
++      find_package(libcudacxx CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(CUB CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      find_package(Thrust CONFIG REQUIRED
++        HINTS ${CUDAToolkit_LIBRARY_DIR}/cmake)
++      thrust_create_target(Thrust HOST CPP DEVICE CUDA)
++      add_library(CCCL::CCCL INTERFACE IMPORTED GLOBAL)
++      target_link_libraries(CCCL::CCCL INTERFACE libcudacxx::libcudacxx CUB::CUB Thrust)
++    endif()
++  endif()
+ endif()
+ 
+ if(FORCE_COLORED_OUTPUT AND (CMAKE_GENERATOR STREQUAL "Ninja") AND
+diff --git a/cmake/Utils.cmake b/cmake/Utils.cmake
+index 8659df7c3..982d4068b 100644
+--- a/cmake/Utils.cmake
++++ b/cmake/Utils.cmake
+@@ -108,12 +108,13 @@ function(xgboost_set_cuda_flags target)
+     target_compile_definitions(${target} PRIVATE -DXGBOOST_USE_NVTX=1)
+   endif()
+ 
++  target_link_libraries(${target}
++    INTERFACE CCCL::CCCL
++    PUBLIC CUDA::cudart_static)
+   target_compile_definitions(${target} PRIVATE -DXGBOOST_USE_CUDA=1)
+   target_include_directories(
+     ${target} PRIVATE
+-    ${xgboost_SOURCE_DIR}/gputreeshap
+-    ${xgboost_SOURCE_DIR}/rabit/include
+-    ${CUDAToolkit_INCLUDE_DIRS})
++    ${xgboost_SOURCE_DIR}/gputreeshap)
+ 
+   if(MSVC)
+     target_compile_options(${target} PRIVATE
+diff --git a/src/common/quantile.h b/src/common/quantile.h
+index 59bc3a4f7..ff3530ab2 100644
+--- a/src/common/quantile.h
++++ b/src/common/quantile.h
+@@ -71,6 +71,14 @@ struct WQSummary {
+          << "value: " << e.value;
+       return os;
+     }
++
++    XGBOOST_DEVICE inline bool operator==(Entry const& rhs) const {
++      return (rmin == rhs.rmin) && (rmax == rhs.rmax)
++             && (wmin == rhs.wmin) && (value == rhs.value);
++    }
++    XGBOOST_DEVICE inline bool operator!=(Entry const& rhs) const {
++      return !operator==(rhs);
++    }
```

Comment:
Do we want to upstream this as well?

Reply:
I don't think we should upstream this patch. In fact, I was about to remove this workaround because it's due to a bug in CCCL. Instead, we should set a filter so that we don't use 2.4.0 or 2.5.0 version of CCCL.

Reply:
Another option is to retain this patch in the Conda recipe, until a new version of CCCL is released.

Reply:
Ah ok. Am now seeing this context in this thread: https://github.com/dmlc/xgboost/pull/10633#discussion_r1690225473

Yep am happy with either option. Will include the other option as a review for comparison/discussion